extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -1,19 +1,173 @@
 import {Services} from '#service';
 
+import {user} from '#utils/log';
+
 import {getElementConfig} from 'extensions/amp-story/1.0/request-utils';
 
+import {assertHttpsUrl} from '../../../src/url';
 import {
   Action,
   ShoppingConfigDataDef,
 } from '../../amp-story/1.0/amp-story-store-service';
 
+/** @const {!Object<!Object<string, !Array<function>>>} */
+
+const productImagesValidation = {
+  'url': [validateRequired, validateURLs],
+  'alt': [validateRequired, validateString],
+};
+
+const aggregateRatingValidation = {
+  'ratingValue': [validateRequired, validateNumber],
+  'reviewCount': [validateRequired, validateNumber],
+  'reviewUrl': [validateRequired, validateURLs],
+};
+
+export const productValidationConfig = {
+  /* Required Attrs */
+  'productUrl': [validateRequired, validateURLs],
+  'productId': [validateRequired, validateString],
+  'productTitle': [validateRequired, validateString],
+  'productBrand': [validateRequired, validateString],
+  'productPrice': [validateRequired, validateNumber],
+  'productImages': [
+    validateRequired,
+    createValidateConfigArray(productImagesValidation),
+  ],
+  'productPriceCurrency': [validateRequired, validateString],
+  /* Optional Attrs */
+  'aggregateRating': [createValidateConfigObject(aggregateRatingValidation)],
+  'productIcon': [validateURLs],
+  'productTagText': [validateString],
+  'ctaText': [validateNumber],
+  'shippingText': [validateNumber],
+};
+
 /**
  * @typedef {{
  *  items: !Array<!ShoppingConfigDataDef>,
  * }}
  */
 let ShoppingConfigResponseDef;
 
+/**
+ * Used for keeping track of intermediary invalid config results within Objects or Arrays of Objects.
+ * @private {boolean}
+ */
+let isValidConfigSection_ = true;
```

Comment:
It's a bit difficult since `createValidateConfigObject` and `createValidateConfigArray` both return a function. 
I could have them return booleans instead of functions.
However, I would then need to do an additional typecheck for `typeof fn === boolean` in
`validationFunctions.forEach((fn) => {`
`fn(configKey, shoppingConfig[configKey]);`
This would seem more difficult to parse than what is already here.
I'm open to ideas, but at this current time this is seems like a more readable approach.

Reply:
@coreymasanto That's what I had initially, but it doesn't work for some reason. 
Even if I return false for `validateConfig`, it would return true in the reducer. 
Hence the workaround. If you think of anything else let me know.

Reply:
`validateConfig()` only sets its `isValidConfig` variable to `false` in the event of an error being thrown from one of the validation functions. So, as with the other validation functions, we need to be throwing errors from your new validation functions instead of returning booleans. If you look at `validateConfig()`, you'll notice that any booleans returned from your new createValidate functions would be completely ignored, so `true` was probably being returned because that's the value to which `isValidConfig` is initialized. Your `isValidConfigSection_` circumvented this issue by making sure that the validation statuses of 'productImages' and 'aggregateRating' are taken into account, but we should be able to do this the correct way.

createValidateConfigObject():
```
return (field, value) => {
  if (!validateConfig()) {
    throw Error(`Value for field '${field}' is not vaild`);
  }
}
```


createValidateConfigArray():
```
return (field, value) => {
  let isValid = true;
  for (const item of value) {
    isValid &&= validateConfig();
  }
  if (!isValid) {
    throw Error(`Value for field '${field}' is not vaild`);
  }
}
```

If this works, please also update `getShoppingConfig()` to use `filter()` instead of `reduce()`, per my other [comment](https://github.com/ampproject/amphtml/pull/37474#discussion_r818923189).

Reply:
Yes, it works, sounds good!